docs/dev/relay_pass_manager.rst

```diff
@@ -0,0 +1,606 @@
+scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
+these modern frameworks are able to conveniently add modules/layers to their
+containers and build up neural network easily.
+
+.. _Sequential: https://pytorch.org/docs/stable/nn.html?highlight=sequential#torch.nn.Sequential
+.. _Block: https://mxnet.incubator.apache.org/_modules/mxnet/gluon/block.html
+
+The design of the Relay pass manager is largely inspired by the the hierarchical
+pass manager used in LLVM and the block style containers used in the popular
+deep learning frameworks. The major goals of the pass manager include:
+
+#) enabling better programmatic orchestration of optimizations. This allows
+   users to flexibly customize and build their own optimization pipelines.
+
+#) providing a user friendly way to debug optimization passes.
+
+#) alleviating developers's from manually and respectively resolving the
+   dependencies between passes.
+
+#) simplifying the implementation of new passes for developers. For example, we
+   allow users to implement a pass in Python and let the pass manager manipulate
+   its execution.
+
+The Design
+----------
+
+We focus on ease of extension for users, making it possible for users to quickly
+add new passes without loss of backward compatibility. The design contains both
+the backend and the frontend. The former implements the main logic of the pass
+manager. The latter provides simple APIs for users to interact with, i.e.
+allowing users to quickly create their own optimization pipelines.
+
+C++ Backend
+~~~~~~~~~~~
+
+We provide a ``PassInfo`` object to contain the basic information needed by
+a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization
+level the pass will be enabled, and ``required`` represents the passes that are
+required to execute a certain pass (see ``include/tvm/relay/transform.h`` for
+more details).
+
+.. code:: c++
+
+    class PassInfoNode : public RelayNode {
+      std::string name;
+      int opt_level;
+      std::vector<std::string> required;
+    };
+
+PassContext
+^^^^^^^^^^^
+
+``PassContext`` carries useful information for an optimization pass. For
+example, it contains the error reporting system so optimization authors can
+provide diagnostics about why an optimization fails. ``PassContext`` is also
+designed to replace the old ``BuildConfig`` which was used to help users
+configure the compilation options, including optimization level and
+required/disabled passes, etc. 
+
+This class is designed as a thread local so that users can conveniently use the
+Python ``with`` syntax to perform optimizations under a certain configuration.
+Examples will be provided later to show how we can use both C++ and Python APIs
+to create a compilation pipeline using pass context.
+
+.. code:: c++
+
+    class PassContextNode : public RelayNode {
+     public:
+      ErrorReporter err_reporter;
+      int opt_level{2};
```

Comment:
Why is opt_level here? Isn't defining it in PassInfoNode sufficient?

Reply:
yeah, each pass is associated with an opt_level. So you will know if a certain pass will be enabled when users provide a opt_level in the configuration context. This is different from what we usually have in a compiler.